File: test/lib/c/03-publish-b2c-qos2-unexpected-pubrel.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <dimq.h>

static int run = -1;

void on_connect(struct dimq *dimq, void *obj, int rc)
{
	if(rc){
		exit(1);
	}
}

void on_message(struct dimq *dimq, void *obj, const struct dimq_message *msg)
{
	if(!strcmp(msg->topic, "quit")){
		run = 0;
		return;
	}
	if(msg->mid != 13423){
		printf("Invalid mid (%d)\n", msg->mid);
		exit(1);
	}
	if(msg->qos != 2){
		printf("Invalid qos (%d)\n", msg->qos);
		exit(1);
	}
	if(strcmp(msg->topic, "pub/qos2/receive")){
		printf("Invalid topic (%s)\n", msg->topic);
		exit(1);
	}
	if(strcmp(msg->payload, "message")){
		printf("Invalid payload (%s)\n", (char *)msg->payload);
		exit(1);
	}
	if(msg->payloadlen != 7){
		printf("Invalid payloadlen (%d)\n", msg->payloadlen);
		exit(1);
	}
	if(msg->retain != false){
		printf("Invalid retain (%d)\n", msg->retain);
		exit(1);
	}

}

int main(int argc, char *argv[])
{
	int rc;
	struct dimq *dimq;

	int port = atoi(argv[1]);

	dimq_lib_init();

	dimq = dimq_new("publish-qos2-test", true, &run);
	if(dimq == NULL){
		return 1;
	}
	dimq_connect_callback_set(dimq, on_connect);
	dimq_message_callback_set(dimq, on_message);
	dimq_message_retry_set(dimq, 5);

	rc = dimq_connect(dimq, "localhost", port, 60);

	while(run == -1){
		rc = dimq_loop(dimq, 300, 1);
		if(rc){
			printf("%d:%s\n", rc, dimq_strerror(rc));
			exit(1);
		}
	}

	dimq_destroy(dimq);
	dimq_lib_cleanup();
	return run;
}

File: test/lib/03-publish-b2c-qos2-unexpected-pubrel.py
#!/usr/bin/env python3

from dimq_test_helper import *

port = dimq_test.get_lib_port()

rc = 1
keepalive = 60
connect_packet = dimq_test.gen_connect("publish-qos2-test", keepalive=keepalive)
connack_packet = dimq_test.gen_connack(rc=0)

disconnect_packet = dimq_test.gen_disconnect()

pubrel_unexpected = dimq_test.gen_pubrel(1000)
pubcomp_unexpected = dimq_test.gen_pubcomp(1000)

mid = 13423
publish_packet = dimq_test.gen_publish("pub/qos2/receive", qos=2, mid=mid, payload="message")
pubrec_packet = dimq_test.gen_pubrec(mid)
pubrel_packet = dimq_test.gen_pubrel(mid)
pubcomp_packet = dimq_test.gen_pubcomp(mid)

publish_quit_packet = dimq_test.gen_publish("quit", qos=0, payload="quit")

sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
sock.settimeout(10)
sock.bind(('', port))
sock.listen(5)

client_args = sys.argv[1:]
env = dict(os.environ)
env['LD_LIBRARY_PATH'] = '../../lib:../../lib/cpp'
try:
    pp = env['PYTHONPATH']
except KeyError:
    pp = ''
env['PYTHONPATH'] = '../../lib/python:'+pp
client = dimq_test.start_client(filename=sys.argv[1].replace('/', '-'), cmd=client_args, env=env, port=port)

try:
    (conn, address) = sock.accept()
    conn.settimeout(10)

    if dimq_test.expect_packet(conn, "connect", connect_packet):
        conn.send(connack_packet)

        conn.send(pubrel_unexpected)
        if dimq_test.expect_packet(conn, "pubcomp", pubcomp_unexpected):

            conn.send(publish_packet)

            if dimq_test.expect_packet(conn, "pubrec", pubrec_packet):
                conn.send(pubrel_packet)

                if dimq_test.expect_packet(conn, "pubcomp", pubcomp_packet):
                    conn.send(publish_quit_packet)
                    rc = 0

    conn.close()
finally:
    for i in range(0, 5):
        if client.returncode != None:
            break
        time.sleep(0.1)

    try:
        client.terminate()
    except OSError:
        pass

    client.wait()
    sock.close()
    if client.returncode != 0:
        exit(1)

exit(rc)
